django/curator/tag_deduplication.py

```diff
@@ -0,0 +1,162 @@
+import abc
+import os
+import re
+from typing import List
+
+import dedupe
+from django.utils.text import slugify
+from taggit.models import Tag
+
+from curator.models import CanonicalTagMapping, CanonicalTag
+
+
+class TagPreprocessing:
+    def preprocess(tag_name: str) -> List[str]:
+        tag_names = re.split(r" and |\;|\,", tag_name)
+        tag_names = [tag_name.strip() for tag_name in tag_names]
+        return tag_names
+
+
+class TagDeduplication(abc.ABC):
+    FIELDS = [{"field": "name", "type": "String"}]
+
+    # Gets the model's most uncertain pairs
+    def uncertain_pairs(self):
+        return self.deduper.uncertain_pairs()
+
+    # Pairs will be two tags.
+    # If the tags refer to the same thing, is_distinct = True
+    # Otherwise, is_distinct should be equal to False
+    def mark_pairs(self, pairs, is_distinct: bool):
+        labelled_examples = {"match": [], "distinct": []}
+
+        if is_distinct:
+            labelled_examples["distinct"] = pairs
+        else:
+            labelled_examples["match"] = pairs
+
+        self.deduper.mark_pairs(labelled_examples)
+
+    def prepare_training_data(self):
+        tags = Tag.objects.all().values()
+        data = {row["id"]: row for i, row in enumerate(tags)}
+        return data
+
+    def console_label(self):
+        dedupe.console_label(self.deduper)
+
+
+# This class is used to help to make the initial canonical list.
+# Besides that, there isn't much other use for this.
+# If the curator can directly provide a small canonical list, even a small one will do, this class will not be needed.
+class TagClustering(TagDeduplication):
+    TRAINING_FILE = "curator/clustering_training.json"
+    CLUSTERING_THRESHOLD = 0.1
```

Comment:
make this parameterizable with defaults defined in django settings (similar to SEARCH_THRESHOLD below)